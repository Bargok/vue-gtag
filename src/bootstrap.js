import { warn, isFn, loadScript } from "./util";
import { getOptions } from "../src/install";
import config from "./api/config";
import query from "./api/query";
import optOut from "./api/opt-out";
import pageTracker from "./page-tracker";

export default function() {
  if (typeof document === "undefined" || typeof window === "undefined") {
    return;
  }

  const {
    enabled,
    globalObjectName,
    config: { id, params = {} },
    pageTrackerEnabled,
    onReady,
    disableScriptLoad
  } = getOptions();

  if (!enabled) {
    optOut();
  }

  window.dataLayer = window.dataLayer || [];
  window[globalObjectName] = function() {
    window.dataLayer.push(arguments);
  };

  query("js", new Date());
  config(params);

  if (pageTrackerEnabled) {
    pageTracker();
  }

  if (disableScriptLoad) {
    return;
  }

  const resource = `https://www.googletagmanager.com/gtag/js?id=${id}`;

  return loadScript(resource)
    .then(() => {
      const library = window[globalObjectName];

      if (isFn(onReady)) {
        onReady(library);
      }

      return library;
    })
    .catch(error => {
      warn("Ops! Something happened and gtag.js couldn't be loaded", error);
      return error;
    });
}
