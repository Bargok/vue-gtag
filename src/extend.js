import { Vue } from "./install";
import query from "./lib/query";
import config from "./lib/config";
import event from "./lib/event";
import pageview from "./lib/pageview";
import screenview from "./lib/screenview";
import customMap from "./lib/custom-map";

export default function() {
  Vue.prototype.$gtag = {
    query,
    config,
    event,
    pageview,
    screenview,
    customMap
  };
}
