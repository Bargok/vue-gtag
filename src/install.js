import { mergeDeep } from "./util";
import extend from "./extend";
import bootstrap from "./bootstrap";

let Vue;
let Router;
let options = {
  enabled: true,
  disableScriptLoad: false,
  bootstrap: true,
  globalObjectName: "gtag",
  pageTrackerTemplate: null,
  onReady: null,
  pageTrackerEnabled: true,
  pageTrackerScreenviewEnabled: false,
  config: {
    id: null,
    params: {
      send_page_view: true
    }
  }
};

export const getOptions = () => options;
export const getVue = () => Vue;
export const getRouter = () => Router;

export function install(_Vue, _options = {}, _Router) {
  Vue = _Vue;
  Router = _Router;
  options = mergeDeep(options, _options);

  extend();

  if (options.bootstrap) {
    bootstrap();
  }
}
