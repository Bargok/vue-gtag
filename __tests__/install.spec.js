import { getVue, getRouter, install, getOptions } from "@/install";
import bootstrap from "@/bootstrap";
import extend from "@/extend";

jest.mock("@/extend");
jest.mock("@/bootstrap");

describe("install", () => {
  it("should have default options", () => {
    expect(getOptions()).toMatchSnapshot();
  });

  it("should return VueRouter", () => {
    install(null, null, { foo: "bar" });
    expect(getRouter()).toEqual({ foo: "bar" });
  });

  it("should return Vue", () => {
    install({ foo: "bar" });
    expect(getVue()).toEqual({ foo: "bar" });
  });

  it("should call extend", () => {
    install();
    expect(extend).toHaveBeenCalled();
  });

  it("should call bootstrap", () => {
    install();
    expect(bootstrap).toHaveBeenCalled();
  });
});
