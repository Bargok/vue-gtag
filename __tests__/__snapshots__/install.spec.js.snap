// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install should have default options 1`] = `
Object {
  "bootstrap": true,
  "config": Object {
    "id": null,
    "params": Object {
      "send_page_view": true,
    },
  },
  "disableScriptLoad": false,
  "enabled": true,
  "globalObjectName": "gtag",
  "onReady": null,
  "pageTrackerEnabled": true,
  "pageTrackerScreenviewEnabled": false,
  "pageTrackerTemplate": null,
}
`;
